src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/LogicalOpUtils.java

```diff
@@ -36,34 +44,31 @@
 import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperator;
 import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalOperatorForLogicalOperator;
 
-import java.util.ArrayList;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
-
 public class LogicalOpUtils
 {
 	/**
 	 * Rewrites the given request operator (with a triple pattern request) into
 	 * a logical plan that uses the local vocabulary of the federation member of
 	 * the request.
 	 */
-	public static LogicalPlan rewriteToUseLocalVocabulary( final LogicalOpRequest<TriplePatternRequest, ?> reqOp ) {
-		final TriplePatternRequest tpReq = reqOp.getRequest();
-		final TriplePattern tp = tpReq.getQueryPattern();
-
+	public static LogicalPlan rewriteToUseLocalVocabulary( final LogicalOpRequest<?, ?> reqOp ) {
 		final FederationMember fm = reqOp.getFederationMember();
-		final VocabularyMapping vm = fm.getVocabularyMapping();
+		if (fm.getVocabularyMapping() == null) { // If no vocabulary mapping, nothing to translate.
+			return new LogicalPlanWithNullaryRootImpl(reqOp);
+		}
+		final SPARQLRequest req = (SPARQLRequest) reqOp.getRequest();
+		final SPARQLGraphPattern p = req.getQueryPattern();
 
-		final SPARQLGraphPattern newP = vm.translateTriplePattern(tp);
 
-		if ( newP.equals(tp) ) {
+		final SPARQLGraphPattern newP = VocabularyMappingUtils.translateGraphPattern(p, fm.getVocabularyMapping());
+		if ( newP.equals(p) ) {
 			return new LogicalPlanWithNullaryRootImpl(reqOp);
```

Comment:
Also here I would prefer that we are not creating a new `LogicalPlan` object but instead return the one that would be given to this function.